### PR TITLE
Add dataset client

### PR DESCRIPTION
**Pull request by alexandersjoberg**
Add dataset client class that interacts with the Peltarion Data API service. The DatasetClient object takes a token and a url, both of which are found on the Dataset overview view on the Peltarion platform. The token and the url are specific to a project. The former is used for authenticating the requester and the latter is the part of the url that stays the same for all requests towards the API service.

In order to upload one or many files to a project, call create_upload_many. This function takes a list of file paths as input and will create a dataset wrapper containing the data that is input to the function. If more than one file path is provided, the files will be appended horizontally to the dataset. Supported file formats are the same as the platform supports from UI, i.e. .csv, .zip, .npy

### CODE REVIEW

File: sidekick/__init__.py [hunk 1]
```diff
@@ -1,13 +1,16 @@
...
 from .dataset import create_dataset, process_image
 from .deployment import Deployment
+from .dataset_client import DatasetClient
 
 __all__ = [
     'Deployment',
+    'DatasetClient',
     'create_dataset',
     'deployment',
+    'dataset_client'
```
Comment (alvaro-garcia-carrasco) [inline sidekick/__init__.py]: Does it need to be public?

File: sidekick/dataset_client.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+        self._headers = {'Authorization': 'Bearer ' + token}
+        self._url = url
+        self._verbose = verbose
+        self._session = requests.Session()
+        self._session.mount('', HTTPAdapter(max_retries=self.MAX_RETRIES))
+        self._session.headers.update({'User-Agent': 'sidekick'})
+
+    def _call_avoin(self, resource: str, verb: str, payload=None) -> str:
+        response = None
+        if self._verbose: print("Calling " + self._url + resource)
```
Comment (alvaro-garcia-carrasco) [inline sidekick/dataset_client.py]: Use `urllib.parse.urljoin` to build the url. It will prevent errors with trailing slashes for instance. Since you also use it three times maybe it is better to define it once with another variable.

File: sidekick/dataset_client.py [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+                data=payload
+            )
+        elif verb == 'get':
+            response = self._session.get(
+                url=self._url + resource,
+                headers=self._headers
+            )
+        else:
+            raise ValueError('HTTP verb ' + verb + ' not supported.')
+
```
Comment (alvaro-garcia-carrasco) [inline sidekick/dataset_client.py]: The Requests package has a nice method `response.raise_for_status()` that will raise and exception if the code is not 200. It could be good to add it here so we catch any issues as soon as possible

File: sidekick/dataset_client.py [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+    def _get_statuses(self, response: str) -> List[UploadStatus]:
+        uploads = json.loads(response)['uploadStatuses']
+        upload_status = []
+        for upload in uploads:
+            upload_status.append(UploadStatus[upload['status']])
+
+        return upload_status
+
+    def _get_content_type(self, filepath: str):
```
Comment (alvaro-garcia-carrasco) [inline sidekick/dataset_client.py]: It could be nice to add the return type as string

File: sidekick/dataset_client.py [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+            content_type = 'text/csv'
+        elif file_extension == '.zip':
+            content_type = 'application/zip'
+        elif file_extension == '.npy':
+            content_type = 'application/npy'
+        else:
+            raise ValueError('Not able to set content-type from file name')
+
+        print("Setting Content-Type to " + content_type) if self._verbose else 0
+        return content_type
```
Comment (alvaro-garcia-carrasco) [inline sidekick/dataset_client.py]: Add empty line

**Comment by agrinh** [comment]
> Add dataset client class that, given a token, url and list of files, creates a dataset wrapper and adds input files horizontally to the wrapper

@alexandersjoberg Can you please develop this a bit? I don't understand what the intention is here.

File: sidekick/dataset_client.py [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+        file_extension = os.path.splitext(filepath)[1]
+        content_type = ""
+        if file_extension == '.csv':
+            content_type = 'text/csv'
+        elif file_extension == '.zip':
+            content_type = 'application/zip'
+        elif file_extension == '.npy':
+            content_type = 'application/npy'
+        else:
+            raise ValueError('Not able to set content-type from file name')
```
Comment (alvaro-garcia-carrasco) [inline sidekick/dataset_client.py]: It could be nice to have information of the supported content types in the exception, what is the file extension given, and which ones are supported

File: sidekick/dataset_client.py [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+        return response.text
+
+    def _create_wrapper(self, name='Sidekick upload', description='Sidekick upload'):
+        self._headers['Content-Type'] = 'application/json'
+        return self._call_avoin('', 'post',
+                                "{\"name\": \"" + name + "\",\"description\": \"" + description + "\"}")
+
+    def _upload_file(self, filepath: str, wrapper_id: str):
+        fileobj = open(filepath, 'rb').read()
+        self._headers['Content-Type'] = self._get_content_type(filepath)
```
Comment (alvaro-garcia-carrasco) [inline sidekick/dataset_client.py]: I think that the `_headers` attribute should not change with every different call because it will make it more prone to silly bugs. I would recommend to use `_token` as attribute instead, and build the header inside the different methods (`_upload_file`, `_create_wrapper`, etc..). Probably using a function to build the header where you pass the token and the content type.

File: sidekick/dataset_client.py [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+
+    def __init__(self, url: str, token: str, verbose=False) -> None:
+        self._headers = {'Authorization': 'Bearer ' + token}
+        self._url = url
+        self._verbose = verbose
+        self._session = requests.Session()
+        self._session.mount('', HTTPAdapter(max_retries=self.MAX_RETRIES))
+        self._session.headers.update({'User-Agent': 'sidekick'})
+
+    def _call_avoin(self, resource: str, verb: str, payload=None) -> str:
```
Comment (alvaro-garcia-carrasco) [inline sidekick/dataset_client.py]: I think that it might be more clear if we had the http call inside the different functions (`_create_wrapper` etc..) but I am up for discussion

File: sidekick/dataset_client.py [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+        print("Response status: " + str(response.status_code) + ", " + response.text) if self._verbose else 0
+        return response.text
+
+    def _create_wrapper(self, name='Sidekick upload', description='Sidekick upload'):
+        self._headers['Content-Type'] = 'application/json'
+        return self._call_avoin('', 'post',
+                                "{\"name\": \"" + name + "\",\"description\": \"" + description + "\"}")
+
+    def _upload_file(self, filepath: str, wrapper_id: str):
+        fileobj = open(filepath, 'rb').read()
```
Comment (alvaro-garcia-carrasco) [inline sidekick/dataset_client.py]: It is a good idea to close the file object or even better using a context manager

File: sidekick/dataset_client.py [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+        return self._call_avoin(f'/{wrapper_id}/uploads', 'get')
+
+    def _complete_upload(self, wrapper_id: str):
+        self._headers['Accept'] = "application/json"
+        return self._call_avoin(f'/{wrapper_id}/upload_complete', 'post')
+
+    def create_upload_many(self, files: List[str], dataset_name='Sidekick upload',
+                           dataset_description='Sidekick upload'):
+        response = self._create_wrapper(dataset_name, dataset_description)
+        wrapper_id = json.loads(response)['datasetWrapperId']
```
Comment (alvaro-garcia-carrasco) [inline sidekick/dataset_client.py]: Depending on how the response object looks like, it could be good to add a data class for it to make an explicit agreement between the data api and sidekick

File: sidekick/dataset_client.py [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+        for file in files:
+            self._upload_file(file, wrapper_id)
+
+        upload_ongoing = True
+        while upload_ongoing:
```
Comment (alvaro-garcia-carrasco) [inline sidekick/dataset_client.py]: I am a bit worried that we might be hitting the dataset_api many times here and that might be a problem. Could it be a good idea to poll it less often (once per second maybe? you could add time.sleep(1)  for instance. Maybe Jakob can give you more info here

File: sidekick/dataset_client.py [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+    def _has_ongoing_uploads(self, response: str):
+        statuses = self._get_statuses(response)
+        return UploadStatus.PROCESSING in statuses or UploadStatus.PENDING in statuses
+
+    def _has_failed_uploads(self, response: str):
+        statuses = self._get_statuses(response)
+        return UploadStatus.FAILED in statuses
+
+    def _get_statuses(self, response: str) -> List[UploadStatus]:
+        uploads = json.loads(response)['uploadStatuses']
```
Comment (alvaro-garcia-carrasco) [inline sidekick/dataset_client.py]: Same as above, it could be nice to have a well defined data structure here

File: sidekick/dataset_client.py [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+    def create_upload_many(self, files: List[str], dataset_name='Sidekick upload',
+                           dataset_description='Sidekick upload'):
+        response = self._create_wrapper(dataset_name, dataset_description)
+        wrapper_id = json.loads(response)['datasetWrapperId']
+        for file in files:
+            self._upload_file(file, wrapper_id)
+
+        upload_ongoing = True
+        while upload_ongoing:
+            response = self._get_upload_status(wrapper_id)
```
Comment (alvaro-garcia-carrasco) [inline sidekick/dataset_client.py]: It could be good to not repeat the variable name `response` since it is also used above

File: sidekick/dataset_client.py [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+            if self._has_failed_uploads(response):
+                raise ValueError('One or more uploads failed. Response: ' + response)
+            upload_ongoing = self._has_ongoing_uploads(response)
```
Comment (alvaro-garcia-carrasco) [inline sidekick/dataset_client.py]: both methods `self._has_failed_uploads(response)` and `self._has_ongoing_uploads(response)` require `statuses = self._get_statuses(response)`, perhaps you could factor that out of the methods so they have a single purpose

File: sidekick/dataset_client.py [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+
+    def _has_failed_uploads(self, response: str):
+        statuses = self._get_statuses(response)
+        return UploadStatus.FAILED in statuses
+
+    def _get_statuses(self, response: str) -> List[UploadStatus]:
+        uploads = json.loads(response)['uploadStatuses']
+        upload_status = []
+        for upload in uploads:
+            upload_status.append(UploadStatus[upload['status']])
```
Comment (alvaro-garcia-carrasco) [inline sidekick/dataset_client.py]: It is a bit more hipster to do a list comprehension: `upload_status = [UploadStatus[upload['status']] for upload in uploads]`

**Comment by alvaro-garcia-carrasco** [comment]
The REAME.md needs to be updated too

File: sidekick/dataset_client.py [hunk 16]
```diff
@@ -0,0 +1,111 @@
...
+        else:
+            raise ValueError('HTTP verb ' + verb + ' not supported.')
+
+        print("Response status: " + str(response.status_code) + ", " + response.text) if self._verbose else 0
+        return response.text
+
+    def _create_wrapper(self, name='Sidekick upload', description='Sidekick upload'):
+        self._headers['Content-Type'] = 'application/json'
+        return self._call_avoin('', 'post',
+                                "{\"name\": \"" + name + "\",\"description\": \"" + description + "\"}")
```
Comment (alvaro-garcia-carrasco) [inline sidekick/dataset_client.py]: It might be clearer to do:
payload = {"name": name, "description": description}
json_payload = json.dumps(payload)

**Comment by alvaro-garcia-carrasco** [comment]
Alexander and I have been working on a dataset_client for Sidekick. The code is now ready to be reviewed, we are missing a README.md update and the build fail with Python3.5 (I would like us to consider maybe to drop supporting it). 

My main issue is that I am not exactely sure of what is the final goal and what is the functionality that needs to be exposed to the user. Maybe we could have a quick meeting to discuss this? 

@alexandersjoberg @lefant @elekaja @agrinh

File: sidekick/dataset_client.py [hunk 17]
```diff
@@ -0,0 +1,123 @@
...
+        self.ongoing = False
+        jobs = self.polling_fn()
+
+        for job in jobs:
+            if job.status is Status.FAILED:
+                self.failed_jobs.append(job.id)
+            elif job.status is Status.SUCCESS:
+                if job.id not in self.successful_jobs:
+                    self.successful_jobs.append(job.id)
+                    print('Job %s successfully saved.' % job.id)
```
Comment (agrinh) [inline sidekick/dataset_client.py]: Remove debugging printout

File: sidekick/dataset_client.py [hunk 18]
```diff
@@ -0,0 +1,123 @@
...
+            raise ValueError(
+                'Valid extensions: %s. Given: %s' % (
+                    supported_extensions, path.suffix)
+            )
+        content_type = self.FILE_EXTENSIONS[path.suffix]
+
+        with open(path, 'rb') as file:
+            data = file.read()
+
+        print('Uploading file %s...' % path)
```
Comment (agrinh) [inline sidekick/dataset_client.py]: Remove debug printout

File: sidekick/dataset_client.py [hunk 19]
```diff
@@ -0,0 +1,123 @@
...
+        response = self._session.post(
+            url=self._url + '%s/upload' % wrapper_id,
+            headers={'Content-Type': content_type},
+            data=data
+        )
+        response.raise_for_status()
+        print('File %s uploaded' % path)
```
Comment (agrinh) [inline sidekick/dataset_client.py]: Remove debug printout

File: examples/dataset_client_example.py [hunk 20]
```diff
@@ -0,0 +1,37 @@
...
+
+        for future in concurrent.futures.as_completed(futures):
+            future.result()
+
+    polling = Polling(partial(client.get_status, wrapper_id))
+    while polling.ongoing:
+        polling.update()
+        if polling.failed_jobs:
+            raise IOError('Failed jobs: %s' % polling.failed_jobs)
+        time.sleep(1)
```
Comment (agrinh) [inline examples/dataset_client_example.py]: I can't help but feel that this approach is a bit harder than it has to be. I don't have a concrete suggestion now, but I'll try to think about this for a little while.

File: sidekick/dataset_client.py [hunk 21]
```diff
@@ -0,0 +1,123 @@
...
+    def to_dict(self):
+        return {'uploadId': self.id, 'status': self.status.name}
+
+    @classmethod
+    def from_dict(cls, data):
+        renamed_data = {
+            'upload_id': data['uploadId'],
+            'status': Status[data['status']],
+        }
+        return cls(**renamed_data)
```
Comment (agrinh) [inline sidekick/dataset_client.py]: You gain very little from this approach, replace this with:

```python
return cls(upload_id=data['uploadId'], status=Status[data['status']])
```

File: sidekick/dataset_client.py [hunk 22]
```diff
@@ -0,0 +1,123 @@
...
+    SUCCESS = 2
+    FAILED = 3
+
+
+class UploadJob:
+    def __init__(self, upload_id: str, status: Status) -> None:
+        self.id = upload_id
+        self.status = status
+
+    def to_dict(self):
```
Comment (agrinh) [inline sidekick/dataset_client.py]: What is the purpose of this method? I see that you use it in the test, can you simply check the values instead?

File: sidekick/dataset_client.py [hunk 23]
```diff
@@ -0,0 +1,123 @@
...
+                'Authorization': 'Bearer %s' % token,
+                'User-Agent': 'sidekick',
+            }
+        )
+
+    def create_wrapper(self, name: str, description: str) -> str:
+        response = self._session.post(
+            url=self._url,
+            headers={'Content-Type': 'application/json'},
+            json={"name": name, "description": description}
```
Comment (agrinh) [inline sidekick/dataset_client.py]: for consistency, s/"/'/

File: tests/test_dataset_client.py [hunk 24]
```diff
@@ -0,0 +1,143 @@
...
+        )
+        output = client.get_status(wrapper_id)
+        assert len(output) == 1
+        assert output[0].status is Status[status]
+        assert output[0].id == upload_id
+
+    @responses.activate
+    def test_file_upload(self, tmp_path):
+        client = DatasetClient(url=self.url, token='')
+        wrapper_id = "b46ce0d238944007bfd8b0877fb4625e"
```
Comment (agrinh) [inline tests/test_dataset_client.py]: for consistency, s/"/'/

File: tests/test_dataset_client.py [hunk 25]
```diff
@@ -0,0 +1,143 @@
...
+            employee_writer.writerow(['Anders', 'Fabian', 'Alex'])
+            employee_writer.writerow(['Martin', 'Daniel', 'Lena'])
+
+        with ZipFile(csv_dataset, 'w') as myzip:
+            myzip.write(csv_dataset)
+
+        assert client.upload_file(csv_dataset, wrapper_id) == upload_id
+
+
+class TestPolling:
```
Comment (agrinh) [inline tests/test_dataset_client.py]: \+\n (http://google.github.io/styleguide/pyguide.html#35-blank-lines)

File: tests/test_dataset_client.py [hunk 26]
```diff
@@ -0,0 +1,143 @@
...
+
+    @responses.activate
+    def test_create_wrapper(self):
+        client = DatasetClient(url=self.url, token='')
+        wrapper_id = "b46ce0d238944007bfd8b0877fb4625e"
+
+        responses.add(
+            method=responses.POST,
+            url=self.url,
+            json={"datasetWrapperId": wrapper_id}
```
Comment (agrinh) [inline tests/test_dataset_client.py]: for consistency, s/"/'/

File: tests/test_dataset_client.py [hunk 27]
```diff
@@ -0,0 +1,143 @@
...
+        )
+        assert wrapper_id == client.create_wrapper(name='', description='')
+
+    @responses.activate
+    def test_complete_upload(self):
+        client = DatasetClient(url=self.url, token='')
+        wrapper_id = "b46ce0d238944007bfd8b0877fb4625e"
```
Comment (agrinh) [inline tests/test_dataset_client.py]: for consistency, s/"/'/

File: tests/test_dataset_client.py [hunk 28]
```diff
@@ -0,0 +1,143 @@
...
+            method=responses.POST,
+            url='%s%s/upload_complete' % (self.url, wrapper_id),
+            status=204,
+        )
+        client.complete_upload(wrapper_id)
+
+    @responses.activate
+    def test_get_status(self):
+        client = DatasetClient(url=self.url, token='')
+        wrapper_id = "b46ce0d238944007bfd8b0877fb4625e"
```
Comment (agrinh) [inline tests/test_dataset_client.py]: for consistency, s/"/'/

File: sidekick/dataset_client.py (hunk 22 above)
Comment (alvaro-garcia-carrasco) [reply sidekick/dataset_client.py]: Yes! good point

File: sidekick/dataset_client.py (hunk 21 above)
Comment (alvaro-garcia-carrasco) [reply sidekick/dataset_client.py]: That is better!

File: sidekick/__init__.py [hunk 29]
```diff
@@ -2,10 +2,12 @@
 
 from . import deployment, encode
 from .dataset import create_dataset, process_image
+from .dataset_client import DatasetClient
```
[the remark this answers is not in the document]
Comment (alvaro-garcia-carrasco) [reply sidekick/__init__.py]: There is already a dataset module

File: sidekick/dataset_client.py [hunk 30]
```diff
@@ -0,0 +1,123 @@
...
+    @classmethod
+    def from_dict(cls, data):
+        renamed_data = {
+            'upload_id': data['uploadId'],
+            'status': Status[data['status']],
+        }
+        return cls(**renamed_data)
+
+
+class DatasetClient:
```
[the remark this answers is not in the document]
Comment (alvaro-garcia-carrasco) [reply sidekick/dataset_client.py]: Good point

File: sidekick/dataset_client.py (hunk 17 above)
Comment (alvaro-garcia-carrasco) [reply sidekick/dataset_client.py]: I think it might be good to leave the printout to inform the user of the progress since uploading the files may take a long time

**Comment by alvaro-garcia-carrasco** [comment]
Thank everyone for the reviews and comments. They helped us to define further the scope of the project. 

The purpose is to have a client with a single method that allows the user to create a dataset and upload a list of files to it. 

There are 4 main steps:
 - Validation: check if files exist and if they have the right extension
 - Create dataset
 - Stage files: Sidekick uploads files to the server
 - Save files: The server saves files
 - Complete upload: The dataset is closed and becomes public to the user in the platform.

If there is an error during the 'Stage files' process, it is retried up to 3 times.

If there is an error in the 'Save files' process, an exception is raised and the user is notified which file produced the error and why. The user must reupload ALL files again. The cleaning up tasks are not responsability of Sidekick. 

It is understood that the ideal behaviour would be for the user to be able to reupload only those files that have issues, but it has been decided to not do that right now.

@alexandersjoberg will update the README.md
@agrinh , @lefant I would appreciate your comments very much. I am exposing as an argument the number of threads to the user, which I would rather not do, but I do not know how to do the tests. The tests are a bit heavy and hard to follow. I would appreciate feedback on them too. Thanks!


I keep getting errors from python35. I would really like to stop supporting it too :-)

File: sidekick/dataset_client.py [hunk 31]
```diff
@@ -0,0 +1,189 @@
...
+    @classmethod
+    def from_dict(cls, data):
+        return cls(
+            upload_id=data['uploadId'],
+            status=Status[data['status']],
+            message=data.get('message'),
+        )
+
+
+class Dataset:
```
Comment (agrinh) [inline sidekick/dataset_client.py]: I would suggest `Uploader` as this name might be confusing given that we have a method called `create_dataset`.

File: README.md [hunk 32]
```diff
@@ -102,6 +102,36 @@ sidekick.create_dataset(
...
+
+client = sidekick.DatasetClient(url='<url>', token='<token>')
+```
+
+This dataset client may now be used to upload one or many files to the Data API service. Uploading files will create a
+dataset to the project that the token is tied to. The Data API consumer could provide `dataset_name` and 
+`dataset_description` to the dataset. If omitted, default name and description will be set to `Sidekick upload` 
+
+```python
+files = ['/path/to/my/file.zip']
```
Comment (agrinh) [inline README.md]: can you show that one could use the `create_dataset` function here? Maybe something like:

```python
dataset_path = '/path/to/dataset.zip'
sidekick.create_dataset(dataset_path, ...)
response = client.upload_files(
    files=[dataset_path],
    dataset_name='My dataset',
    dataset_description='My description'
)
```
